File: min_select/min_select.js
#!/usr/bin/node
// To the extent possible under law, the person who associated CC0 with
// this project has waived all copyright and related or neighboring rights
// to this project.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
//

function _swap(a,i,j) {
  let t = a[i];
  a[i] = a[j];
  a[j] = t;
}

function bubble_sort(a,s,n) {
  let m = -1;

  for (let i=s; i<(s+n); i++) {
    m = s+n - (i-s);
    for (let j=(s+1); j<m; j++) {
      if (a[j-1] > a[j]) { _swap(a, j-1, j); }
    }
  }
}

function k_select_recur(a, s, n, kl) {
  let v=-1, b=-1, e=-1, m=-1;
  let g = Math.floor(n/5);

  if (g==0) {
    if (n>0) {
      bubble_sort(a,s,n);
      _swap(a,s,s+kl);
    }
    return a[s];
  }

  for (let i=0; i<g; i++) {
    bubble_sort(a, s+5*i, s+5);
    _swap(a, s+i, s+(5*i)+2);
  }

  k_select_recur(a,s,g,Math.floor(n/10));

  v = a[s];

  for (b=(s+1), e=(s+n-1); (e-b)>0; ) {
    if (a[b] > v) { _swap(a,b,e); e--; }
    else          { b++; }
  }

  if (a[b] > v) { b--; }
  _swap(a,s,b);

  m = b-s;
  if (m == kl) {
    _swap(a,s,s+kl);
    return a[0];
  }

  else if (kl < m) {
    k_select_recur(a,s,m,kl);
  }

  else {
    // m = b-s -> s = b-m -> s+m = b
    k_select_recur(a, s + (m+1), n - (m+1), kl - (m+1) );
    _swap(a, s, b+1);
  }

  return a[0];
}

function k_select(a, kl) {
  return k_select_recur(a, 0, a.length, kl);
}



if (typeof module !== "undefined") {
  module.exports["_select"] = k_select_recur;
  module.exports["select"] = k_select;
}

if ((typeof require !== "undefined") &&
    (require.main === module)) {

  function _S(a) { return JSON.stringify(a); }

  function icmp(u,v) {
    if (u<v) { return -1; }
    if (u>v) { return 1; }
    return 0;
  }

  function _main(argv) {

    let n = 37;

    let a = [],
        b = [],
        _orig = [];

    let n_it = 10;

    let pass = true;

    /*
    let A = [0, 1, 13, 5, 2, 7,  7, 7 ];
    let B = [];
    for (let i=0; i<A.length; i++) { B.push(A[i]); }
    B.sort(icmp);
    console.log("select 7 from", _S(A), _S(B));
    let v_A = k_select_recur(A, 0, A.length, 7);
    console.log(">>>", v_A);
    return;
    */

    /*
    let A = [3,9,7,4,0];
    console.log("select 3 from", A, A.sort(icmp));
    let v_A = k_select_recur(A, 0, A.length, 3);
    console.log(">>>", v_A);
    return;
    */

    for (let n=20; n<45; n++) {

      for (let it=0; it<n_it; it++) {

        _orig = [];
        b = [];

        for (let i=0; i<n; i++) { _orig.push( Math.floor(Math.random()*2*n) ); }
        for (let i=0; i<n; i++) { b.push(_orig[i]); }
        b.sort( (_u,_v) => { if (_u<_v) { return -1; } if (_u>_v) { return 1; } return 0; } );

        for (let k=0; k<n; k++) {
          a = [];
          for (let i=0; i<n; i++) { a.push( _orig[i] ); }

          let v_a = k_select_recur(a, 0, n, k);

          console.log("it:", it, "k:", k, "n:", n, "val_a:", v_a, "val_b:", b[k]);

          if (v_a != b[k]) { pass = false; console.log("NOOOOO", it, k, a, b, _orig); }
        }
      }

      if (pass) { console.log("pass"); }
      else { console.log("FAIL"); }

    }

  }

  _main(process.argv.slice(1));
}

File: min_select/min_select.c
// To the extent possible under law, the person who associated CC0 with
// this project has waived all copyright and related or neighboring rights
// to this project.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <stdint.h>

int swap(int *a, int *b) {
  int t;
  t = *a;
  *a = *b;
  *b = t;
  return 0;
}

void print_a_flat(int *a, int n) {
  int i;
  printf("[%i]:", n);
  for (i=0; i<n; i++) {
    printf(" %i", a[i]);
  }
  printf("\n");
}

int simpsort(int *a, int n) {
  int i, j, k;
  for (i=0; i<n; i++) {
    for (j=1; j<(n-i); j++) {
      if (a[j-1] > a[j]) { swap(a+j-1, a+j); }
    }
  }
  return 0;
}

int bubble_sort(int *a, int n) {
  int i, j, k;
  for (i=0; i<n; i++) {
    for (j=1; j<(n-i); j++) {
      if (a[j-1] > a[j]) { swap(a+j-1, a+j); }
    }
  }
  return 0;
}

int select_k_large(int *a, int n, int kl) {
  int i, v, b, e;
  int g = n/5;

  // base case, simple sort, putting
  // median value in 0 position,
  // then return
  //
  if (!g) {
    if (n) {
      bubble_sort(a,n); 
      swap(a, a+kl);
    }
    return a[0];
  }

  // sort blocks of 5
  // collect median of each 5 block into
  //   beginning of array.
  //
  for (i=0; i<g; i++) {
    simpsort(a+5*i,5);
    swap(a+i, a+(5*i)+2);
  }

  // recur
  // We want the median (halfway element) of
  // the remaining n/5 elements, so n/10
  //
  select_k_large(a,g, n/10);

  // a[0] is our pivot element now
  //
  v = a[0];

  // collect elements smaller than pivot
  // at the beginning and larger than the pivot
  // at the end
  //
  for (b=1, e=(n-1); (e-b)>0; ) {
    if (a[b] > v) { swap(a+b, a+e); e--; }
    else          { b++; }
  }

  // put pivot element in the middle
  //
  if (a[b] > v) { b--; }
  swap(a, a+b);

  // if we've found our kl largest element,
  // put it in position 0 and return
  //
  if (b == kl) {
    swap(a, a+kl);
    return a[0];
  }

  // else if kl smaller, recur on left side
  //
  else if (kl < b) {
    select_k_large(a,b,kl);
  }

  // otherwise recur on right side
  //
  else {
    select_k_large(a+b+1,n-b-1,kl-b-1);
    swap(a+b+1, a);
  }

  return a[0];
}

void fisher_yates_shuffle(int *a, int n) {
  int i, j, k;
  for (i=0; i<(n-1); i++) {
    k = i + (rand()%(n-i));
    swap(a+i, a+k);
  }
}

void print_a(int *a, int n) {
  int i;

  for (i=0; i<n; i++) {
    printf("[%i]: %i\n", i, a[i]);
  }
}

int icmp(const void *a, const void *b) {
  int *ia, *ib;
  ia = (int *)a;
  ib = (int *)b;

  if ( (*ia) < (*ib) ) { return -1; }
  if ( (*ia) > (*ib) ) { return  1; }
  return 0;
}

int _experiment(int n, int n_it) {
  int i, j, k, it = 0;
  int *A = NULL, *B = NULL, *_orig = NULL;

  _orig = (int *)malloc(sizeof(int)*n);
  A = (int *)malloc(sizeof(int)*n);
  B = (int *)malloc(sizeof(int)*n);

  for (it=0; it<n_it; it++) {

    for (i=0; i<n; i++) { _orig[i] = rand()%(2*n); }
    for (i=0; i<n; i++) { B[i] = _orig[i]; }
    qsort(B, n, sizeof(int), icmp);

    for (k=0; k<n; k++) {

      for (i=0; i<n; i++) { A[i] = _orig[i]; }

      select_k_large(A, n, k);

      if (A[0] != B[k]) { return -1; }

      printf("\nit%i.k%i:", it, k);
      print_a_flat(A, n);
      printf("it%i.k%i:", it, k);
      print_a_flat(B, n);
    }

  }

  free(A);
  free(B);
  free(_orig);

  return 0;
}

int selection_algorithm(int *a, int n, int idx) {
  select_k_large(a,n,idx);
  return a[0];
}

int ___main(int argc, char **argv) {
  int i, n = 37, *a = NULL,
      val, idx;
  int seed = 0;

  if (argc > 1) { seed = atoi(argv[1]); }

  srand(seed);

  a = (int *)malloc(sizeof(int)*n);
  for (i=0; i<n; i++) { a[i] = rand()%(2*n); }

  idx = rand()%n;
  val = selection_algorithm(a,n,idx);

  printf("a[%i]:", n);
  for (i=0; i<n; i++) { printf(" %i", a[i]); }
  printf("\n%i'th value: %i\n", idx, val);

  free(a);
}

int main(int argc, char **argv) {
  int n = 10, r = -1;
  if (argc > 1) {
    n = atoi(argv[1]);
    if (n < 0) { n = 10; }
  }

  r = _experiment(n, 10);

  printf("got: %i\n", r);

}

int _main(int argc, char **argv) {
  int *A = NULL, *B = NULL;
  int i, j, k, n = 26;

  k = 12;
  if (argc > 1) {
    k = atoi(argv[1]);
    if (k < 0) { k = 0; }
    if (k >= n) { k = n-1; }
  }

  A = (int *)malloc(sizeof(int)*n);
  B = (int *)malloc(sizeof(int)*n);

  for (i=0; i<n; i++) { A[i] = i; }


  fisher_yates_shuffle(A,n);

  for (i=0; i<n; i++) { B[i] = A[i]; }

  select_k_large(A,n, k);

  //simpsort(A,n);

  print_a(A,n);

  free(A);
}
